In [2]:
%reload_ext sql

In [3]:
%sql postgresql://appdev@data/appdev

'Connected: appdev@appdev'

In [8]:
%sql SELECT * FROM circuits LIMIT 20;

20 rows affected.


[(1, 'albert_park', 'Albert Park Grand Prix Circuit', 'Melbourne', 'Australia', -37.8497, 144.968, 10, 'http://en.wikipedia.org/wiki/Melbourne_Grand_Prix_Circuit', '(144.968,-37.8497)'),
 (2, 'sepang', 'Sepang International Circuit', 'Kuala Lumpur', 'Malaysia', 2.76083, 101.738, None, 'http://en.wikipedia.org/wiki/Sepang_International_Circuit', '(101.738,2.76083)'),
 (3, 'bahrain', 'Bahrain International Circuit', 'Sakhir', 'Bahrain', 26.0325, 50.5106, None, 'http://en.wikipedia.org/wiki/Bahrain_International_Circuit', '(50.5106,26.0325)'),
 (4, 'catalunya', 'Circuit de Barcelona-Catalunya', 'Montmeló', 'Spain', 41.57, 2.26111, None, 'http://en.wikipedia.org/wiki/Circuit_de_Barcelona-Catalunya', '(2.26111,41.57)'),
 (5, 'istanbul', 'Istanbul Park', 'Istanbul', 'Turkey', 40.9517, 29.405, None, 'http://en.wikipedia.org/wiki/Istanbul_Park', '(29.405,40.9517)'),
 (6, 'monaco', 'Circuit de Monaco', 'Monte-Carlo', 'Monaco', 43.7347, 7.42056, None, 'http://en.wikipedia.org/wiki/Circuit_de_Monaco', '(7.42056,43.7347)'),
 (7, 'villeneuve', 'Circuit Gilles Villeneuve', 'Montreal', 'Canada', 45.5, -73.5228, None, 'http://en.wikipedia.org/wiki/Circuit_Gilles_Villeneuve', '(-73.5228,45.5)'),
 (8, 'magny_cours', 'Circuit de Nevers Magny-Cours', 'Magny Cours', 'France', 46.8642, 3.16361, None, 'http://en.wikipedia.org/wiki/Circuit_de_Nevers_Magny-Cours', '(3.16361,46.8642)'),
 (9, 'silverstone', 'Silverstone Circuit', 'Silverstone', 'UK', 52.0786, -1.01694, None, 'http://en.wikipedia.org/wiki/Silverstone_Circuit', '(-1.01694,52.0786)'),
 (10, 'hockenheimring', 'Hockenheimring', 'Hockenheim', 'Germany', 49.3278, 8.56583, None, 'http://en.wikipedia.org/wiki/Hockenheimring', '(8.56583,49.3278)'),
 (11, 'hungaroring', 'Hungaroring', 'Budapest', 'Hungary', 47.5789, 19.2486, None, 'http://en.wikipedia.org/wiki/Hungaroring', '(19.2486,47.5789)'),
 (12, 'valencia', 'Valencia Street Circuit', 'Valencia', 'Spain', 39.4589, -0.331667, None, 'http://en.wikipedia.org/wiki/Valencia_Street_Circuit', '(-0.331667,39.4589)'),
 (13, 'spa', 'Circuit de Spa-Francorchamps', 'Spa', 'Belgium', 50.4372, 5.97139, None, 'http://en.wikipedia.org/wiki/Circuit_de_Spa-Francorchamps', '(5.97139,50.4372)'),
 (14, 'monza', 'Autodromo Nazionale di Monza', 'Monza', 'Italy', 45.6156, 9.28111, None, 'http://en.wikipedia.org/wiki/Autodromo_Nazionale_Monza', '(9.28111,45.6156)'),
 (15, 'marina_bay', 'Marina Bay Street Circuit', 'Marina Bay', 'Singapore', 1.2914, 103.864, None, 'http://en.wikipedia.org/wiki/Marina_Bay_Street_Circuit', '(103.864,1.2914)'),
 (16, 'fuji', 'Fuji Speedway', 'Oyama', 'Japan', 35.3717, 138.927, None, 'http://en.wikipedia.org/wiki/Fuji_Speedway', '(138.927,35.3717)'),
 (17, 'shanghai', 'Shanghai International Circuit', 'Shanghai', 'China', 31.3389, 121.22, None, 'http://en.wikipedia.org/wiki/Shanghai_International_Circuit', '(121.22,31.3389)'),
 (18, 'interlagos', 'Autódromo José Carlos Pace', 'São Paulo', 'Brazil', -23.7036, -46.6997, None, 'http://en.wikipedia.org/wiki/Aut%C3%B3dromo_Jos%C3%A9_Carlos_Pace', '(-46.6997,-23.7036)'),
 (19, 'indianapolis', 'Indianapolis Motor Speedway', 'Indianapolis', 'USA', 39.795, -86.2347, None, 'http://en.wikipedia.org/wiki/Indianapolis_Motor_Speedway', '(-86.2347,39.795)'),
 (20, 'nurburgring', 'Nürburgring', 'Nürburg', 'Germany', 50.3356, 6.9475, None, 'http://en.wikipedia.org/wiki/N%C3%BCrburgring', '(6.9475,50.3356)')]

_1. On the table circuits report:_
- What type of indices exists for the table and why they are of that type (and not some other type)

In [37]:
%sql select * from pg_indexes where tablename = 'circuits';

3 rows affected.


[('f1db', 'circuits', 'idx_17102_url', None, 'CREATE UNIQUE INDEX idx_17102_url ON circuits USING btree (url)'),
 ('f1db', 'circuits', 'circuits_position_idx', None, 'CREATE INDEX circuits_position_idx ON circuits USING gist ("position")'),
 ('f1db', 'circuits', 'idx_17102_primary', None, 'CREATE UNIQUE INDEX idx_17102_primary ON circuits USING btree (circuitid)')]

##### _idx_17102_url(Btree)_ and _idx_17102_primary(Btree)_
Btree indexes are valuable on the most common data types such as text, numbers and timestamps.

##### _circuits_position_idx (GiST)_
GiSt can work for different data structures including spatial data wherein it is usually stored as coordinates. 


- The amount of space each index takes up

In [34]:
%sql SELECT pg_size_pretty(pg_relation_size('idx_17102_url'));

1 rows affected.


[('16 kB',)]

In [31]:
%sql SELECT pg_size_pretty(pg_relation_size('circuits_position_idx'));

1 rows affected.


[('8192 bytes',)]

In [28]:
%sql SELECT pg_size_pretty(pg_relation_size('idx_17102_primary'));

1 rows affected.


[('16 kB',)]

_2. We are talent scouts looking to win over some of the best new drivers there are. But we don't want them too old! Write a query that finds the winner of all the races, but only if they are younger than 38 years. The query should give return the date, driver surname, driver age, track time in milliseconds, race name and circuit name for all races._

In [67]:
%%sql
SELECT races.date as "Date", drivers.surname as "Driver's Surname", date_part('year', age('2018-03-19',dob)) as "Driver's Age", races.time as "Track Time(ms)", races.name as "Name", circuits.name as "Circuit Name" 
FROM results
JOIN drivers USING (driverid)
JOIN races USING (raceId)
JOIN circuits USING (circuitid)
WHERE results.rank = 1
AND date_part('year', age('2018-03-19',dob)) < 38;

154 rows affected.


[(datetime.date(2008, 3, 16), 'Kovalainen', 36.0, datetime.time(4, 30), 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2008, 4, 6), 'Kovalainen', 36.0, datetime.time(11, 30), 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2008, 8, 24), 'Massa', 36.0, datetime.time(12, 0), 'European Grand Prix', 'Valencia Street Circuit'),
 (datetime.date(2008, 10, 12), 'Massa', 36.0, datetime.time(4, 30), 'Japanese Grand Prix', 'Fuji Speedway'),
 (datetime.date(2008, 10, 19), 'Hamilton', 33.0, datetime.time(7, 0), 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2008, 11, 2), 'Massa', 36.0, datetime.time(17, 0), 'Brazilian Grand Prix', 'Autódromo José Carlos Pace'),
 (datetime.date(2007, 4, 8), 'Hamilton', 33.0, datetime.time(7, 0), 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2007, 4, 15), 'Massa', 36.0, datetime.time(11, 30), 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2007, 5, 13), 'Massa', 36.0, datetime.time(12, 0), 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya'),
 (datetime.date(2007, 5, 27), 'Alonso', 36.0, datetime.time(12, 0), 'Monaco Grand Prix', 'Circuit de Monaco'),
 (datetime.date(2007, 6, 10), 'Alonso', 36.0, datetime.time(17, 0), 'Canadian Grand Prix', 'Circuit Gilles Villeneuve'),
 (datetime.date(2007, 7, 1), 'Massa', 36.0, datetime.time(12, 0), 'French Grand Prix', 'Circuit de Nevers Magny-Cours'),
 (datetime.date(2007, 7, 22), 'Massa', 36.0, datetime.time(12, 0), 'European Grand Prix', 'Nürburgring'),
 (datetime.date(2007, 9, 9), 'Alonso', 36.0, datetime.time(12, 0), 'Italian Grand Prix', 'Autodromo Nazionale di Monza'),
 (datetime.date(2007, 9, 16), 'Massa', 36.0, datetime.time(12, 0), 'Belgian Grand Prix', 'Circuit de Spa-Francorchamps'),
 (datetime.date(2007, 9, 30), 'Hamilton', 33.0, datetime.time(4, 30), 'Japanese Grand Prix', 'Fuji Speedway'),
 (datetime.date(2007, 10, 7), 'Massa', 36.0, datetime.time(6, 0), 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2006, 3, 12), 'Rosberg', 32.0, datetime.time(14, 30), 'Bahrain Grand Prix', 'Bahrain International Circuit'),
 (datetime.date(2006, 3, 19), 'Alonso', 36.0, datetime.time(15, 0), 'Malaysian Grand Prix', 'Sepang International Circuit'),
 (datetime.date(2006, 4, 23), 'Alonso', 36.0, datetime.time(14, 0), 'San Marino Grand Prix', 'Autodromo Enzo e Dino Ferrari'),
 (datetime.date(2006, 5, 14), 'Massa', 36.0, datetime.time(14, 0), 'Spanish Grand Prix', 'Circuit de Barcelona-Catalunya'),
 (datetime.date(2006, 6, 11), 'Alonso', 36.0, datetime.time(12, 0), 'British Grand Prix', 'Silverstone Circuit'),
 (datetime.date(2006, 8, 6), 'Massa', 36.0, datetime.time(14, 0), 'Hungarian Grand Prix', 'Hungaroring'),
 (datetime.date(2006, 10, 1), 'Alonso', 36.0, datetime.time(14, 0), 'Chinese Grand Prix', 'Shanghai International Circuit'),
 (datetime.date(2006, 10, 8), 'Alonso', 36.0, datetime.time(14, 0), 'Japanese Grand Prix', 'Suzuka Circuit'),
 (datetime.date(2005, 3, 6), 'Alonso', 36.0, datetime.time(14, 0), 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2005, 5, 29), 'Alonso', 36.0, datetime.time(14, 0), 'European Grand Prix', 'Nürburgring'),
 (datetime.date(2009, 3, 29), 'Rosberg', 32.0, datetime.time(6, 0), 'Australian Grand Prix', 'Albert Park Grand Prix Circuit'),
 (datetime.date(2009, 5, 24), 'Massa', 36.0, datetime.time(12, 0), 'Monaco Grand Prix', 'Circuit de Monaco'),
 (datetime.date(2009, 6, 21), 'Vettel', 30.0, datetime.time(12, 0), 'British Grand Prix', 'Silverstone Circuit'),
 (datetime.date(2009, 7, 12), 'Alonso', 36.0, datetime.time(12, 0), 'German Grand Prix', 'Nürburgring'),
 (datetime.date(2009, 8, 23), 'Glock', 36.0, datetime.time(12, 0), 'European Grand Prix', 'Valencia Street Circuit'),
 (datetime.date(2009, 8, 30), 'Vettel', 30.0, datetime.time(12, 0), 'Belgian Grand Prix', 'Circuit de Spa-Francorchamps'),
 (datetime.date(2009, 9, 13), 'Sutil', 35.0, datetime.time(12, 0)


_3. Describe the query using EXPLAIN ANALYZE with at least 5 lines of text. Answer at least the following:_

In [83]:
%%sql EXPLAIN ANALYZE
SELECT races.date as "Date", drivers.surname as "Driver's Surname", date_part('year', age('2018-03-19',dob)) as "Driver's Age", races.time as "Track Time(ms)", races.name as "Name", circuits.name as "Circuit Name" 
FROM results
JOIN drivers USING (driverid)
JOIN races USING (raceId)
JOIN circuits USING (circuitid)
WHERE results.rank = 1
AND date_part('year', age('2018-03-19',dob)) < 38;

22 rows affected.


[('Hash Join  (cost=77.92..792.94 rows=86 width=66) (actual time=1.108..7.877 rows=154 loops=1)',),
 ('  Hash Cond: (races.circuitid = circuits.circuitid)',),
 ('  ->  Hash Join  (cost=74.28..787.57 rows=86 width=50) (actual time=1.056..7.601 rows=154 loops=1)',),
 ('        Hash Cond: (results.raceid = races.raceid)',),
 ('        ->  Hash Join  (cost=34.32..746.53 rows=86 width=19) (actual time=0.546..7.026 rows=154 loops=1)',),
 ('              Hash Cond: (results.driverid = drivers.driverid)',),
 ('              ->  Seq Scan on results  (cost=0.00..708.96 rows=257 width=16) (actual time=0.006..6.412 rows=257 loops=1)',),
 ('                    Filter: (rank = 1)',),
 ('                    Rows Removed by Filter: 23420',),
 ('              ->  Hash  (cost=30.82..30.82 rows=280 width=19) (actual time=0.532..0.532 rows=60 loops=1)',),
 ('                    Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                    ->  Seq Scan on drivers  (cost=0.00..30.82 rows=280 width=19) (actual time=0.009..0.513 rows=60 loops=1)',),
 ("                          Filter: (date_part('year'::text, age('2018-03-19 00:00:00+00'::timestamp with time zone, (dob)::timestamp with time zone)) < '38'::double precision)",),
 ('                          Rows Removed by Filter: 781',),
 ('        ->  Hash  (cost=27.76..27.76 rows=976 width=47) (actual time=0.503..0.503 rows=976 loops=1)',),
 ('              Buckets: 1024  Batches: 1  Memory Usage: 83kB',),
 ('              ->  Seq Scan on races  (cost=0.00..27.76 rows=976 width=47) (actual time=0.003..0.247 rows=976 loops=1)',),
 ('  ->  Hash  (cost=2.73..2.73 rows=73 width=28) (actual time=0.042..0.042 rows=73 loops=1)',),
 ('        Buckets: 1024  Batches: 1  Memory Usage: 13kB',),
 ('        ->  Seq Scan on circuits  (cost=0.00..2.73 rows=73 width=28) (actual time=0.007..0.021 rows=73 loops=1)',),
 ('Planning time: 0.410 ms',),
 ('Execution time: 7.961 ms',)]

_EXPLAIN ANALYZE returns a detailed query plan, wherein it shows the actions executed represented by '->'. The query was executed and timing information was defined as values in actual time, rows and loops. So, the first three actions shows the hash conditions such as this:   Hash Join action: (races.circuitid = circuits.circuitid) - meaning this action was executed 1 time(the loops value), that it returned 154 rows, and that actual time was 0... The hash nodes/action also includes the number of buckets and batches, together with the corresponding memory usage._ 
    
- How many calls are you making? <br>
    _There are 10 actions taken on this query._
    
- How long does it take to perform the query?<br>
    _It took 7.961ms to perform the query._
    



_4. Create a materialized view of your query. Using EXPLAIN ANALYZE try to query the view. Write at least 5 lines of text explaining what's going on and why the query execution time changed._

In [81]:
%%sql CREATE VIEW winners AS
SELECT races.date as "Date", drivers.surname as "Driver's Surname", date_part('year', age('2018-03-19',dob)) as "Driver's Age", races.time as "Track Time(ms)", races.name as "Name", circuits.name as "Circuit Name" 
FROM results
JOIN drivers USING (driverid)
JOIN races USING (raceId)
JOIN circuits USING (circuitid)
WHERE results.rank = 1
AND date_part('year', age('2018-03-19',dob)) < 38;

(psycopg2.ProgrammingError) relation "winners" already exists
 [SQL: 'CREATE VIEW winners AS\nSELECT races.date as "Date", drivers.surname as "Driver\'s Surname", date_part(\'year\', age(\'2018-03-19\',dob)) as "Driver\'s Age", races.time as "Track Time(ms)", races.name as "Name", circuits.name as "Circuit Name" \nFROM results\nJOIN drivers USING (driverid)\nJOIN races USING (raceId)\nJOIN circuits USING (circuitid)\nWHERE results.rank = 1\nAND date_part(\'year\', age(\'2018-03-19\',dob)) < 38;']


In [86]:
%sql EXPLAIN ANALYZE SELECT * FROM winners;

22 rows affected.


[('Hash Join  (cost=77.92..792.94 rows=86 width=66) (actual time=0.769..3.500 rows=154 loops=1)',),
 ('  Hash Cond: (races.circuitid = circuits.circuitid)',),
 ('  ->  Hash Join  (cost=74.28..787.57 rows=86 width=50) (actual time=0.728..3.367 rows=154 loops=1)',),
 ('        Hash Cond: (results.raceid = races.raceid)',),
 ('        ->  Hash Join  (cost=34.32..746.53 rows=86 width=19) (actual time=0.350..2.949 rows=154 loops=1)',),
 ('              Hash Cond: (results.driverid = drivers.driverid)',),
 ('              ->  Seq Scan on results  (cost=0.00..708.96 rows=257 width=16) (actual time=0.005..2.554 rows=257 loops=1)',),
 ('                    Filter: (rank = 1)',),
 ('                    Rows Removed by Filter: 23420',),
 ('              ->  Hash  (cost=30.82..30.82 rows=280 width=19) (actual time=0.333..0.333 rows=60 loops=1)',),
 ('                    Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                    ->  Seq Scan on drivers  (cost=0.00..30.82 rows=280 width=19) (actual time=0.009..0.319 rows=60 loops=1)',),
 ("                          Filter: (date_part('year'::text, age('2018-03-19 00:00:00+00'::timestamp with time zone, (dob)::timestamp with time zone)) < '38'::double precision)",),
 ('                          Rows Removed by Filter: 781',),
 ('        ->  Hash  (cost=27.76..27.76 rows=976 width=47) (actual time=0.372..0.372 rows=976 loops=1)',),
 ('              Buckets: 1024  Batches: 1  Memory Usage: 83kB',),
 ('              ->  Seq Scan on races  (cost=0.00..27.76 rows=976 width=47) (actual time=0.003..0.190 rows=976 loops=1)',),
 ('  ->  Hash  (cost=2.73..2.73 rows=73 width=28) (actual time=0.032..0.032 rows=73 loops=1)',),
 ('        Buckets: 1024  Batches: 1  Memory Usage: 13kB',),
 ('        ->  Seq Scan on circuits  (cost=0.00..2.73 rows=73 width=28) (actual time=0.007..0.017 rows=73 loops=1)',),
 ('Planning time: 0.367 ms',),
 ('Execution time: 3.567 ms',)]

In [89]:
%sql CREATE MATERIALIZED VIEW winners_cache AS SELECT * FROM winners;

154 rows affected.


[]

In [90]:
%sql EXPLAIN ANALYZE SELECT * FROM winners_cache;

3 rows affected.


[('Seq Scan on winners_cache  (cost=0.00..10.50 rows=50 width=1568) (actual time=0.007..0.028 rows=154 loops=1)',),
 ('Planning time: 0.086 ms',),
 ('Execution time: 0.050 ms',)]

_Materialized views are cached views, and that's why it only calls one action. One time execution, that returned 154 rows in 0 actual time. Though this will not update automatically whenever the data has been updated. It just stores static snapshots and saves the time. It only ran once and storing the result, it requires a refresh and might take some time. It may violate normal form, but this is just a query. It will be really useful when the database is used often in reading than updating._